duck/call.py

```diff
@@ -0,0 +1,171 @@
+
+    def has_exact_args(self, *args, **kwargs):
+        """Returns ``True`` if the call was made with exactly the provided
+        arguments and keyword arguments, and ``False`` otherwise.
+
+        If provided, positional arguments must start at ``0`` and be in order.
+
+        Comparisons are performed with ``==``, not ``is``.
+        """
+        # There must be the same number of arguments.
+        if len(args) != len(self.args):
+            return False
+
+        # Arguments must be identical.
+        for idx, actual in enumerate(args):
+            if actual != self.args[idx]:
+                return False
+
+        # There must be the same number of keyword arguments.
+        if len(kwargs) != len(self.kwargs):
+            return False
+
+        # They must be identical.
+        for k, v in kwargs.items():
+            if k not in self.kwargs or self.kwargs[k] != v:
+                return False
+
+        return True
+
+    def assert_has_args(self, *args, **kwargs):
+        """Calls the ``has_args`` method, and raises ``AssertionError``
+        if ``False`` is returned.
+        """
+        errors = []
+
+        # Determine if the given arguments are valid.
+        if not self.has_args(*args):
+            errors.append('Given args: {} != Valid args: {}'.format(
+                args,
+                self.args,
+            ))
+
+        # Determine if the given keyword arguments are valid.
+        if not self.has_args(**kwargs):
+            errors.append('Given kwargs: {} != Valid kwargs: {}'.format(
+                kwargs,
+                self.kwargs,
+            ))
+
+        if errors:
+            raise AssertionError('\r\n'.join(errors))
+
+        return True
+
+    def assert_has_exact_args(self, *args, **kwargs):
+        """Calls the ``has_exact_args`` method, and raises ``AssertionError``
+        if ``False`` is returned.
+
+        In order to return useful debugging information, it post-processes
+        similar to ``has_exact_args``, but with additional information.
+        """
+        errors = []
+
+        # Determine if the full set of (kw)arguments is valid.
+        if not self.has_exact_args(*args, **kwargs):
+
+            # There must be the same number of arguments.
+            if len(args) != len(self.args):
+                errors.append('Given args: {} != Actual args: {}'.format(
+                    args,
+                    self.args,
+                ))
+            else:
+                # Arguments must be identical.
+                args_error = False
+                for idx, actual in enumerate(args):
+                    if actual != self.args[idx]:
+                        args_error = True
+
+                if args_error:
+                    errors.append('Given args: {} != Actual args: {}'.format(
+                        args,
+                        self.args,
+                    ))
+
+            # There must be the same number of keyword arguments.
```

Comment:
I would recommend checking the sets of keys here rather than just the length:

```python
expected_keys = {k for k in kwargs.keys()}
actual_keys = {k for k in self.kwargs.keys()}
if expected_keys != actual_keys:
   ...
```

This will isolate errata from mismatched _keys_ from errata from mismatched _values_, which will be useful.